### PR TITLE
Improve performance of exec-hash

### LINKED ISSUE #3734: Improve exec hash performance
Today we include hash for binaries in some events, most notably the process execution events. A couple of performance improvements have been suggested:
1. more efficient caching: today we do cache the calculated hashes to avoid recalculating subsequent executions. but the cache is scoped to the mount namespace, for example same container executed many times will still recalculate the hash. we can improve it by sharing hash calculations across namespaces.
2. use a more efficient hash function. it appears there's an alternative library that could easily win performance gain.

In the future, we should also look into redesigning the event processing pipeline to be more asynchronous.

### LINKED ISSUE #3745: exec-hash doesn't set the sha256 arg on the first call
I have detected that on the first event with a given filepath, exec-hash calculates the hash but doesn't set the sha256 arg.

`sudo ./dist/tracee -s comm=uname -e sched_process_exec -o option:exec-hash --capabilities add=cap_sys_ptrace -o json`

#### First call on the host

```
host pid: 3199690, pid: 3199690, mntns: 4026531841
device: 66308, inode: 5506453
sourceFilePath: /proc/1/root/usr/bin/uname
capturedFileID: host:/usr/bin/uname
filePath: /usr/bin/uname, castedSourceFileCtime: 1697222770429185913

{"timestamp":1701882096151912175,"threadStartTime":1701882096150820564,"processorId":0,"processId":3199690,"cgroupId":37075,"threadId":3199690,"parentProcessId":3174263,"hostProcessId":3199690,"hostThreadId":3199690,"hostParentProcessId":3174263,"userId":1000,"mountNamespace":4026531841,"pidNamespace":4026531836,"processName":"uname","executable":{"path":""},"hostName":"hb","containerId":"","container":{},"kubernetes":{},"eventId":"713","eventName":"sched_process_exec","matchedPolicies":[""],"argsNum":12,"returnValue":0,"syscall":"execve","stackAddresses":null,"contextFlags":{"containerStarted":false,"isCompat":false},"threadEntityId":4078181535,"processEntityId":4078181535,"parentEntityId":293296475,"args":[{"name":"cmdpath","type":"const char*","value":"/usr/bin/uname"},{"name":"pathname","type":"const char*","value":"/usr/bin/uname"},{"name":"dev","type":"dev_t","value":271581188},{"name":"inode","type":"unsigned long","value":5506453},{"name":"ctime","type":"unsigned long","value":1697222770429185913},{"name":"inode_mode","type":"umode_t","value":33261},{"name":"interpreter_pathname","type":"const char*","value":null},{"name":"interpreter_dev","type":"dev_t","value":null},{"name":"interpreter_inode","type":"unsigned long","value":null},{"name":"interpreter_ctime","type":"unsigned long","value":null},{"name":"argv","type":"const char**","value":["uname"]},{"name":"interp","type":"const char*","value":"/usr/bin/uname"},{"name":"stdin_type","type":"string","value":"S_IFCHR"},{"name":"stdin_path","type":"char*","value":"/dev/pts/5"},{"name":"invoked_from_kernel","type":"int","value":0},{"name":"env","type":"const char**","value":null},{"name":"sha256","type":"const char*","value":""}]}
```

#### Second call on the host

```
host pid: 3199741, pid: 3199741, mntns: 4026531841
device: 66308, inode: 5506453
sourceFilePath: /proc/1/root/usr/bin/uname
capturedFileID: host:/usr/bin/uname
filePath: /usr/bin/uname, castedSourceFileCtime: 1697222770429185913

{"timestamp":1701882099909166161,"threadStartTime":1701882099908218741,"processorId":10,"processId":3199741,"cgroupId":37075,"threadId":3199741,"parentProcessId":3174263,"hostProcessId":3199741,"hostThreadId":3199741,"hostParentProcessId":3174263,"userId":1000,"mountNamespace":4026531841,"pidNamespace":4026531836,"processName":"uname","executable":{"path":""},"hostName":"hb","containerId":"","container":{},"kubernetes":{},"eventId":"713","eventName":"sched_process_exec","matchedPolicies":[""],"argsNum":12,"returnValue":0,"syscall":"execve","stackAddresses":null,"contextFlags":{"containerStarted":false,"isCompat":false},"threadEntityId":1245852069,"processEntityId":1245852069,"parentEntityId":293296475,"args":[{"name":"cmdpath","type":"const char*","value":"/usr/bin/uname"},{"name":"pathname","type":"const char*","value":"/usr/bin/uname"},{"name":"dev","type":"dev_t","value":271581188},{"name":"inode","type":"unsigned long","value":5506453},{"name":"ctime","type":"unsigned long","value":1697222770429185913},{"name":"inode_mode","type":"umode_t","value":33261},{"name":"interpreter_pathname","type":"const char*","value":null},{"name":"interpreter_dev","type":"dev_t","value":null},{"name":"interpreter_inode","type":"unsigned long","value":null},{"name":"interpreter_ctime","type":"unsigned long","value":null},{"name":"argv","type":"const char**","value":["uname"]},{"name":"interp","type":"const char*","value":"/usr/bin/uname"},{"name":"stdin_type","type":"string","value":"S_IFCHR"},{"name":"stdin_path","type":"char*","value":"/dev/pts/5"},{"name":"invoked_from_kernel","type":"int","value":0},{"name":"env","type":"const char**","value":null},{"name":"sha256","type":"const char*","value":"d97b71d4cb4cc402d04c6a4623885fcf337687279eb24640f42ee94960ad6489"}]}
```

_Originally posted by @geyslan in https://github.com/aquasecurity/tracee/issues/3734#issuecomment-1843302069_

### PR DESCRIPTION
**Pull request by NDStrahilevitz**
### 1. Explain what the PR does

c2ebbb290 **feat(exechash): optimize hash by reusing buffer**
9f9e5a44d **feat(ebpf): expand exec-hash output option**
eb17fc6fc **feat(ebpf): optimize computeFileHash**
db51e0b0e **fix(ebpf): fix getFileHash()**
c469486e5 **fix(bucketscache): GetBucket was not thread safe**
041bcff5e **chore(bucketscache): add benchmark**

### 2. Explain how to test it

Added tests to confirm parity with previous hash functionality.
Can be manually tested by using:
`tracee -o option:exec-hash={inode|dev-inode|digest-inode}`
where you may choose which hash option you want, with varying degrees of accuracy and performance.

Close #3734
Close #3745

### CODE REVIEW

File: pkg/cmd/flags/output_test.go [hunk 1]
```diff
@@ -325,14 +325,14 @@ func TestPrepareOutput(t *testing.T) {
 			},
 		},
 		{
-			testName:    "option exec-hash",
-			outputSlice: []string{"option:exec-hash"},
+			testName:    "option exec-hash=pathname",
```
Comment (NDStrahilevitz) [inline pkg/cmd/flags/output_test.go:328]: todo: fix this test

File: docs/docs/flags/output.1.md [hunk 2]
```diff
@@ -48,7 +48,10 @@ Other options:
   - **exec-env**: When tracing execve/execveat, show the environment variables that were used for execution.
   - **relative-time**: Use relative timestamp instead of wall timestamp for events.
   - **exec-hash**: When tracing some file related events, show the file hash (sha256).
-    - Affected events: sched_process_exec, shared_object_loaded
+    - Affected events: *sched_process_exec*, *shared_object_loaded*
+    - **pathname** option recalculates the file hash if the inode's creation time (ctime) differs, which can occur in different namespaces even for identical pathnames.
```
Comment (AlonZivony) [inline docs/docs/flags/output.1.md:52]: This documentation is not updated to the new options

File: docs/man/output.1 [hunk 3]
```diff
@@ -90,7 +90,21 @@ timestamp for events.
 file hash (sha256).
 .RS 2
 .IP \[bu] 2
-Affected events: sched_process_exec, shared_object_loaded
+Affected events: \f[I]sched_process_exec\f[R],
+\f[I]shared_object_loaded\f[R]
+.IP \[bu] 2
+\f[B]inode\f[R] option recalculates the file hash if the inode\[cq]s
+creation time (ctime) differs, which can occur in different namespaces
+even for identical pathnames.
```
Comment (AlonZivony) [inline docs/man/output.1:98]: Maybe add a note that this should reduce performance in the case of deploying the same image for multiple containers?
It should be clear that this is the preferred method for these cases if the container enrichment is not available

File: pkg/cmd/flags/output_test.go
Comment (AlonZivony) [inline pkg/cmd/flags/output_test.go:1]: I think having a test case for wrong option is also important

File: pkg/exechash/cache.go
Comment (AlonZivony) [inline pkg/exechash/cache.go:1]: About the package name - the name `exec-hash` is legacy, and it is no longer precise.
Now we have hashes for more events, so I propose to call call the package `filehash` or something similar.

**Review by AlonZivony — CHANGES_REQUESTED** [review]

File: docs/docs/flags/output.1.md (hunk 2 above)
Comment (NDStrahilevitz) [reply docs/docs/flags/output.1.md:52]: Thanks for catching.
Unfortunately `man` pages aren't working for me for whatever reason so it's easier to forget.

File: pkg/cmd/flags/output.go [hunk 4]
```diff
@@ -113,6 +111,29 @@ func setOption(cfg *config.OutputConfig, option string, newBinary bool) error {
 	case "sort-events":
 		cfg.EventsSorting = true
 	default:
+		if strings.HasPrefix(option, "exec-hash") {
```
[the remark this answers is not in the document]
Comment (NDStrahilevitz) [reply pkg/cmd/flags/output.go:114]: True, will fix.

File: docs/docs/flags/output.1.md [hunk 5]
```diff
@@ -11,7 +11,7 @@ tracee **\-\-output** - Control how and where output is printed
 
 ## SYNOPSIS
 
-tracee **\-\-output** <format[:file,...]\> | gotemplate=template[:file,...] | forward:url | webhook:url | option:{stack-addresses,exec-env,relative-time,exec-hash,parse-arguments,parse-arguments-fds,sort-events} ...
+tracee **\-\-output** <format[:file,...]\> | gotemplate=template[:file,...] | forward:url | webhook:url | option:{stack-addresses,exec-env,relative-time,exec-hash={pathname,dev-inode,digest-inode},parse-arguments,parse-arguments-fds,sort-events} ...
```
Comment (AlonZivony) [inline docs/docs/flags/output.1.md:14]: pathname->inode

File: pkg/cmd/flags/output.go [hunk 6]
```diff
@@ -113,6 +111,29 @@ func setOption(cfg *config.OutputConfig, option string, newBinary bool) error {
...
+				case "inode":
+					cfg.CalcHashes = config.CalcHashesInode
+				case "dev-inode":
+					cfg.CalcHashes = config.CalcHashesDevInode
+				case "digest-inode":
+					cfg.CalcHashes = config.CalcHashesDigestInode
+				default:
+					goto invalidOption
+				}
+			}
```
Comment (AlonZivony) [inline pkg/cmd/flags/output.go:131]:
```Go
} else {
    goto invalidOption
}
```

**Review by AlonZivony — CHANGES_REQUESTED** [review]
Small things

File: docs/docs/flags/output.1.md (hunk 5 above)
Comment (rafaeldtinoco) [reply docs/docs/flags/output.1.md:14]: if all of them are `XXX-inode`, should we remove `inode` ?

File: pkg/filehash/key.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	filePath string
+	mountNS  int
+
+	device      uint32
+	inode       uint64
+	ctime       int64
+	imageDigest string
+}
+
+func NewKey(filepath string, mountNS int, opts ...func(*Key)) Key {
```
Comment (rafaeldtinoco) [inline pkg/filehash/key.go:22]: New key creates bleh bleh.

File: pkg/filehash/key.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+	k := Key{filePath: filepath, mountNS: mountNS}
+	for _, o := range opts {
+		o(&k)
+	}
+	return k
+}
+
+func WithDevice(device uint32) func(*Key) {
```
Comment (rafaeldtinoco) [inline pkg/filehash/key.go:30]: WithDevice bleh bleh bleh

File: pkg/filehash/key.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+	return func(k *Key) {
+		k.device = device
+	}
+}
+
+func WithInode(ino uint64, ctime int64) func(*Key) {
```
Comment (rafaeldtinoco) [inline pkg/filehash/key.go:36]: You got the point.

**Review by rafaeldtinoco — APPROVED** [review]
LGTM. It is missing some public methods documentation lines, but other than that (and minors from Alon) I think we're good to merge. Also, nice job in using SIMD optimized SHA package, it has multiple x86 flavors (SSE, ...) and arm as well (I wonder if it supports SVE extensions in arm, would make it super fast).

File: docs/docs/flags/output.1.md [hunk 10]
```diff
@@ -48,7 +48,10 @@ Other options:
   - **exec-env**: When tracing execve/execveat, show the environment variables that were used for execution.
   - **relative-time**: Use relative timestamp instead of wall timestamp for events.
   - **exec-hash**: When tracing some file related events, show the file hash (sha256).
-    - Affected events: sched_process_exec, shared_object_loaded
+    - Affected events: *sched_process_exec*, *shared_object_loaded*
+    - **inode** option recalculates the file hash if the inode's creation time (ctime) differs, which can occur in different namespaces even for identical inode. This option is performant, but not recommended and should only be used if container enrichment can't be enabled for digest-inode, and if performance is preffered over correctness.
```
Comment (yanivagman) [inline docs/docs/flags/output.1.md:52]: Ctime is inode change time, not creation time

File: docs/docs/flags/output.1.md [hunk 11]
```diff
@@ -48,7 +48,10 @@ Other options:
...
+    - **dev-inode** option generally offers better performance compared to the **inode** option, as it bypasses the need for recalculation by associating the creation time (ctime) with the device (dev) and inode pair. It's recommended if correctness is preffered over performance without container enrichment.
```
Comment (yanivagman) [inline docs/docs/flags/output.1.md:53]: You say this option offers better performance, while on the end of this sentence it says that this option is more for correctness, and not performance. This is confusing

**Review by yanivagman — COMMENTED** [review]
I didn't do a full review, but left a couple of comments about the doc file

**Pull request merged**